include/jemalloc/internal/thread_event.h

```diff
@@ -0,0 +1,175 @@
+
+JEMALLOC_ALWAYS_INLINE void
+thread_event_adjust_thresholds_helper(tsd_t *tsd, uint64_t last_event,
+    uint64_t wait) {
+	assert(tsd_reentrancy_level_get(tsd) == 0);
+	assert(last_event == tsd_thread_allocated_last_event_get(tsd));
+	assert(wait > 0);
+	uint64_t threshold = last_event + wait;
+	tsd_thread_allocated_threshold_set(tsd, threshold);
+	if (likely(last_event < threshold &&
+	    threshold <= THREAD_ALLOCATED_THRESHOLD_FAST_MAX)) {
+		tsd_thread_allocated_threshold_fast_set(tsd, threshold);
+	} else {
+		/*
+		 * threshold either overflowed or exceeded
+		 * THREAD_ALLOCATED_THRESHOLD_FAST_MAX.  Zero
+		 * thread_allocated_threshold_fast so that
+		 * thread_event_fast() will always trigger.
+		 */
+		tsd_thread_allocated_threshold_fast_set(tsd, 0);
+	}
+}
+
+JEMALLOC_ALWAYS_INLINE bool
+thread_event(tsd_t *tsd, size_t usize) {
+	if (tsd_reentrancy_level_get(tsd) > 0) {
+		return false;
+	}
+
+	uint64_t thread_allocated = tsd_thread_allocated_get(tsd);
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	if (likely(usize < tsd_thread_allocated_threshold_get(tsd) -
+	    thread_allocated)) {
+		return false;
+	}
+
+	/*
+	 * An event is triggered, so bulk transfer the accumulated allocation
+	 * bytes to all wait times, and obtain the new threshold.
+	 */
+
+	/*
+	 * (thread_allocated + usize - last_event) may overflow int64_t, and
+	 * in such cases the int64_t cast may result in a negative number, but
+	 * when bulk_transfer is subtracted from the int64_t wait times, the
+	 * subtraction will also overflow, and the result is still correct.
+	 */
+	int64_t bulk_transfer = (int64_t)(thread_allocated + usize -
+	    tsd_thread_allocated_last_event_get(tsd));
+
+	/*
+	 * In case none of the events has its condition on, the
+	 * THREAD_EVENT_WAIT_MAX initialization will automatically delay the
+	 * next event as much as possible.
+	 */
+	uint64_t wait = THREAD_EVENT_WAIT_MAX;
+
+#define W(event, condition)						\
+	if (condition) {						\
+		int64_t event_wait = tsd_##event##_event_wait_get(tsd);	\
+		event_wait -= bulk_transfer;				\
+		tsd_##event##_event_wait_set(tsd, event_wait);		\
+		/* If there is a single event, then the remaining     */\
+		/* wait	time won't be positive, and we rely on a      */\
+		/* thread_event_register() call later to properly set */\
+		/* the next threshold; if there are multiple events,  */\
+		/* then here we can get the minimum remaining wait    */\
+		/* time to the next already registered event.	      */\
+		if (event_wait > 0 && (uint64_t)event_wait < wait) {	\
+			wait = (uint64_t)event_wait;			\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_WAITS
+#undef W
+
+	uint64_t last_event = thread_allocated + usize;
+	tsd_thread_allocated_last_event_set(tsd, last_event);
+	thread_event_adjust_thresholds_helper(tsd, last_event, wait);
+	return true;
+}
+
+#define W(event, condition)						\
+JEMALLOC_ALWAYS_INLINE void						\
+thread_##event##_event_register(tsd_t *tsd, uint64_t wait) {		\
```

Comment:
Not a fan on _event_register being written in macro. Let's simply pass in the appropriate pointers into this function and have one version instead? Also this should happen every time after an event fires, so no need to make it inline really.

I slightly prefer _event_update or _refresh. I'd imagine the "register" should only happen once per event type on init, and we keep updating it.

Reply:
Makes sense. On inlining: the only reason why I'm inlining it is that it's in the header file. Perhaps shall I instead create an `extern` header file, an `inline` header file (for the fast path function), and a source file for the slow path functions?

Reply:
re: splitting headers -- so far the rule is around these lines: if the header is simply / small enough, then a single header containing everything is fine. Otherwise we split into the _extern, _struct, _types etc. as needed. In this case I feel a single head is still alright, after we move the slow path into the source file.

Reply:
The reason for this is dependency breaking; the multiple headers/module broke circular dependencies in the build graph (the idea being to refactor them away over time). Ideally we shouldn't be adding any more cycles though, so I don't know that we need them for new code.